Identify DLLs from the stage

johnwparent opened the pull request:
Relocating DLL and their corresponding libs from the stage currently has a pitfall, where post installation, the paths inside import libraries to their corresponding DLLs is no longer valid, as it points to the DLL as it existed at link time, which is the stage. We cannot presume where the DLL will end up in the install tree vs the stage, and we need valid "rpaths" at test time, so we are forced to marshall libs and their dlls post install.

Since we cannot 1:1 map a location in an install tree to a location in the stage, we need some other mechanism for aligning these two binaries. Previously we simply used the symbols exported by each, but since import libraries and DLLs can export different sets of symbols, despite the fact they are a import lib/dll pair, we cannot use that approach.

Instead, inject the stage path inside the dll via a resource file, so when we go to do relocation post install, we can simply query the stage dll path from the import library, search for a dll with a corresponding entry in the string table, and know we've identified a pair. 

This PR:

Creates a resource descriptor file (res) and injects the current path to the dll (stage time) into the string table with the identifier as a custom, unique to Spack, "spackresource" type added into the first pass link line that creates the initial dll.

scheibelp commented on src/ld.cxx:
```diff
@@ -99,3 +113,45 @@ DWORD LdInvocation::InvokeToolchain() {
+    rc_executor.Execute();
+    DWORD const err_code = rc_executor.Join();
+    if (err_code != 0) {
+        return std::string();
```
(question) is it not catastrophic if this fails? Should it raise an exception?

scheibelp reviewed (commented):
A few preliminary questions/requests

johnwparent committed:
Relocating DLL and their corresponding libs from the stage currently has a pitfall, where post installation, the paths inside import libraries to their corresponding DLLs is no longer valid, as it points to the DLL as it existed at link time, which is the stage. We cannot presume where the DLL will end up in the install tree vs the stage, and we need valid "rpaths" at test time, so we are forced to marshall libs and their dlls post install.

Since we cannot 1:1 map a location in an install tree to a location in the stage, we need some other mechanism for aligning these two binaries. Previously we simply used the symbols exported by each, but since import libraries and DLLs can export different sets of symbols, despite the fact they are a import lib/dll pair, we cannot use that approach.

Instead, inject the stage path inside the dll via a resource file, so when we go to do relocation post install, we can simply query the stage dll path from the import library, search for a dll with a corresponding entry in the string table, and know we've identified a pair.

This PR:

Creates a resource descriptor file (res) and injects the current path to the dll (stage time) into the string table with the identifier "59673" (randomly selected number, it just needs to be any consistent identifier) it into the first pass link line that creates the initial dll.
The identifier needs to be a numeric, and this is a random one I chose.

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
remaining typo

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent commented:
Rebased on #32

johnwparent commented on src/ld.cxx:
```diff
@@ -99,3 +113,45 @@ DWORD LdInvocation::InvokeToolchain() {
+    // This string table ID is completely arbitrary, HOWEVER
+    // Spack Core relies on this specific value
+    // If it is changed here, it must be changed in Spack Core
+    const std::string string_table_id = "    59673 ";
```
Closing the loop on this thread since I think most of the followup took place on slack/in meetings:

We can create an arbitrary resource, and did. 
We created `spackresource` as a type, and inject it into the dlls, so there will be no id clashes.

scheibelp commented on src/ld.cxx:
```diff
@@ -95,3 +159,48 @@ DWORD LdInvocation::InvokeToolchain() {
+    }
+
+    ExecuteCommand rc_executor("rc",
+                               {"/fo" + res_file_name + " " + rc_file_name});
```
It looks like these files are being created in the CWD. I think as a follow-up they should be made in a temp dir.
